keps/sig-node/3008-cri-class-based-resources/README.md

```diff
@@ -0,0 +1,1257 @@
+the available memory bandwidth.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+This is only the first step in getting class-based resources supported in
+Kubernetes. Important pieces like resource assignment via pod spec, resource
+status, resource disovery and permission control are [non-goals](#non-goals)
+not solved here.
+These aspects
+are briefly discussed in [future work](#future-work). The risk in this sort of
+piecemeal approach is finding devil in the details, resulting in inconsistent
+and/or crippled and/or cumbersome end result. However, there is a lot of
+experience in extending the API and understanding which sort of solutions are
+functional and practical.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+- User assigning container to “unauthorized” class, causing interference and
+  access to unwanted set/amount of resources. This will be addressed in future
+  KEP introducing permission controls.
+- Confusion: user tries to assign container to RDT class but RDT has not been
+  enabled on system(s). This will be addressed by future KEP(s) introducing
+  resource discovery and status.
+- Keeping client (kubelet) and runtime in sync wrt to available classes. Will
+  be addressed in future KEP about resource discovery.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+Configuration and management of the class resources is fully handled by the
+underlying container runtime and is invisible to kubelet. An error to the CRI
+client is returned if the specified class is not available.
+
+### CRI protocol
+
+The following additions to the CRI protocol are suggested.
+
+The `ContainerConfig` message will be supplemented with new `class_resources`
+field, providing per-container setting for class resources.
+
+
+```diff
+ message ContainerConfig {
+
+ ...
+     // Configuration specific to Linux containers.
+     LinuxContainerConfig linux = 15;
+     // Configuration specific to Windows containers.
+     WindowsContainerConfig windows = 16;
++
++    // Configuration of class resources.
++    ContainerClassResources class_resources = 17;
```

Comment:
with the work on InPlace update of CPU and Memory, will classes be needed to be in-place updated in future? Will the need for this change the proposed API surface?

Reply:
This is a good note that I already touched in one of the previous comments. Yes, I think we should prepare for this in the CRI API now (even if it wouldn't work because of runtime limitations at present). I haven't updated the KEP yet to reflect this as I want to check the details

Reply:
I think it's reasonable to leave support out until we have a valid use case. Kubelet has a lot of knowledge of information for InPlace to be enabled, and that's against the design decisions here. I think it's reasonable as developers to say: if folks want in place upgrade support, it has to have a first-class QoS field

Reply:
@vinaykul is the InPlace update today has a single setting whether resource can be applied without container restart? Or this can/should be per-resource. If we will start updating QoS in-place, those classes likely will not need a restart while memory/CPU will.

Reply:
In the proposal, the Mutable field in the QoSResourceInfo CRI message (and corresponding field in the QoSResourceInfo type in the K8s API) is intended to indicate whether a specific QoS resource can be updated in-place, without restart. @vinaykul any comments on this?